### PR TITLE
UpdateActivityOptions, continue

### PR DESCRIPTION
**Pull request by ychebotarev**
## What changed?
* Change the logic. Now we regenerate retry task for every call.
* Add `RegenerateRetryTask`. Pls check the logic.
*  Add replication for all recent fields
* Propagate activity `stamp` as a part of activity task
* "Reject" activity start in case of stamp mismatch from history service
* Update "standby" timer task processed to take into account `stamp` for retry tasks
(and do nothing for timeout tasks)

## Why?
Continue workfing on UpdateActivityOptions API

Known issues: need more functional tests.

### CODE REVIEW

File: common/serviceerror/activity_stamp_mismatch.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+func (e *ActivityStampMismatch) Status() *status.Status {
+	if e.st != nil {
+		return e.st
+	}
+
+	st := status.New(codes.InvalidArgument, e.Message)
+	st, _ = st.WithDetails(
+		&errordetails.TaskAlreadyStartedFailure{},
```
Comment (yycptt) [inline common/serviceerror/activity_stamp_mismatch.go:60]: ActivityStampMismatchFailure

File: service/history/timer_queue_active_task_executor.go
Comment (yycptt) [inline service/history/timer_queue_active_task_executor.go:1]: We need to talk about this comment in the first PR: https://github.com/temporalio/temporal/pull/6657/files#r1809737126

File: service/history/timer_queue_active_task_executor.go [hunk 2]
```diff
@@ -574,6 +574,7 @@ func (t *timerQueueActiveTaskExecutor) executeActivityRetryTimerTask(
 		ScheduleToStartTimeout: durationpb.New(scheduleToStartTimeout),
 		Clock:                  vclock.NewVectorClock(t.shardContext.GetClusterMetadata().GetClusterID(), t.shardContext.GetShardID(), task.TaskID),
 		VersionDirective:       directive,
+		Stamp:                  task.Stamp,
```
Comment (yycptt) [inline service/history/timer_queue_active_task_executor.go:577]: we need the same thing in transfer activity task.

File: service/history/timer_queue_standby_task_executor.go [hunk 3]
```diff
@@ -321,6 +321,11 @@ func (t *timerQueueStandbyTaskExecutor) executeActivityRetryTimerTask(
 			return nil, nil
 		}
 
+		if activityInfo.Stamp != task.Stamp {
```
Comment (yycptt) [inline service/history/timer_queue_standby_task_executor.go:324]: standby activity retry timer may also push activity to matching, we also need to specify stamp there

File: service/history/workflow/mutable_state_impl.go [hunk 4]
```diff
@@ -4912,6 +4911,35 @@ func (ms *MutableStateImpl) RecordLastActivityStarted(ai *persistencespb.Activit
 	})
 }
 
+func (ms *MutableStateImpl) RegenerateActivityRetryTask(ai *persistencespb.ActivityInfo) error {
```
Comment (yycptt) [inline service/history/workflow/mutable_state_impl.go:4914]: nit: probably rename this to something else, it's more than just generating a task.

File: service/history/replication/raw_task_converter.go [hunk 5]
```diff
@@ -193,6 +193,9 @@ func convertActivityStateReplicationTask(
 						LastStartedRedirectCounter: activityInfo.GetUseWorkflowBuildIdInfo().GetLastRedirectCounter(),
 						BaseExecutionInfo:          persistence.CopyBaseWorkflowInfo(mutableState.GetBaseWorkflowInfo()),
 						VersionHistory:             versionhistory.CopyVersionHistory(currentVersionHistory),
+						FirstScheduledTime:         activityInfo.FirstScheduledTime,
+						LastAttemptCompleteTime:    activityInfo.LastAttemptCompleteTime,
+						Stamp:                      activityInfo.Stamp,
```
Comment (yycptt) [inline service/history/replication/raw_task_converter.go:198]: Also need to update standby cluster logic to apply those fields (executable_activity_state_task.go, activity_state_replicator.go etc.) Can probably do all replication related logic in a separate PR.

File: common/serviceerror/activity_stamp_mismatch.go (hunk 1 above)
Comment (ychebotarev) [reply common/serviceerror/activity_stamp_mismatch.go:60]: done

File: service/history/replication/raw_task_converter.go (hunk 5 above)
Comment (ychebotarev) [reply service/history/replication/raw_task_converter.go:198]: I updated SyncActivityRequest, ActivitySyncInfo, etc. May be even more then needed (information in SyncActivityRequest is duplicated in ActivitySyncInfo, but also ActivitySyncInfo is created then from SyncActivityRequest). Please check if all code path are covered.

File: common/serviceerror/activity_stamp_mismatch.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	"fmt"
+
+	"go.temporal.io/server/api/errordetails/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type (
+	// ActivityStampMismatch represents actity related task that has diffent stamp.
+	ActivityStampMismatch struct {
```
Comment (bergundy) [inline common/serviceerror/activity_stamp_mismatch.go:35]: You don't need this new type. Just use NotFound error with a good error message.

File: common/serviceerror/activity_stamp_mismatch.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+func (e *ActivityStampMismatch) Error() string {
+	return e.Message
+}
+
+func (e *ActivityStampMismatch) Status() *status.Status {
+	if e.st != nil {
+		return e.st
+	}
+
+	st := status.New(codes.InvalidArgument, e.Message)
```
Comment (bergundy) [inline common/serviceerror/activity_stamp_mismatch.go:58]: Arguably this would be `NotFound` instead of `InvalidArgument`. This is what we typically do to reject tasks.

File: proto/internal/temporal/server/api/historyservice/v1/request_response.proto [hunk 8]
```diff
@@ -274,6 +274,7 @@ message RecordActivityTaskStartedRequest {
     temporal.api.workflowservice.v1.PollActivityTaskQueueRequest poll_request = 6;
     temporal.server.api.clock.v1.VectorClock clock = 7;
     temporal.server.api.taskqueue.v1.BuildIdRedirectInfo build_id_redirect_info = 8;
+    int32 stamp = 9;
```
Comment (bergundy) [inline proto/internal/temporal/server/api/historyservice/v1/request_response.proto:277]: Please ensure all proto fields have comments on them. We've been putting more of an effort to keep the documentation standards for the codebase.

Comment (bergundy) [reply proto/internal/temporal/server/api/historyservice/v1/request_response.proto:277]: I see you were good about this below, must have missed this one.

File: proto/internal/temporal/server/api/persistence/v1/tasks.proto [hunk 9]
```diff
@@ -50,6 +50,9 @@ message TaskInfo {
     // TaskVersionDirective, which is unversioned.)
     // Deprecated. No need to store this for new versioning [cleanup-old-wv]
     temporal.server.api.taskqueue.v1.TaskVersionDirective version_directive = 8;
+
+    // Stamp field allows to differentiate between different instances of the same task
```
Comment (bergundy) [inline proto/internal/temporal/server/api/persistence/v1/tasks.proto:54]:
```suggestion
    // A version marker that is used to drop tasks that are no longer relevant to their corresponding state machine. 
```

File: proto/internal/temporal/server/api/replication/v1/message.proto [hunk 10]
```diff
@@ -165,6 +165,12 @@ message SyncActivityTaskAttributes {
     string last_started_build_id = 16;
     // workflows redirect_counter value when this activity started last time
     int64 last_started_redirect_counter = 17;
+    // First time the activity was scheduled.
```
Comment (bergundy) [inline proto/internal/temporal/server/api/replication/v1/message.proto:168]:
```suggestion
    // The first time the activity was scheduled.
```

File: proto/internal/temporal/server/api/replication/v1/message.proto [hunk 11]
```diff
@@ -165,6 +165,12 @@ message SyncActivityTaskAttributes {
...
+    google.protobuf.Timestamp first_scheduled_time = 18;
+    // Last time an activity failure was recorded by the server.
```
Comment (bergundy) [inline proto/internal/temporal/server/api/replication/v1/message.proto:170]:
```suggestion
    // The last time an activity attempt completion was recorded by the server.
```

File: proto/internal/temporal/server/api/replication/v1/message.proto [hunk 12]
```diff
@@ -165,6 +165,12 @@ message SyncActivityTaskAttributes {
...
+    google.protobuf.Timestamp last_attempt_complete_time = 19;
+    // stamp represent "version" of activity during its lifetime. It can be changed with Activity API
```
Comment (bergundy) [inline proto/internal/temporal/server/api/replication/v1/message.proto:172]: I had a suggestion above for this doc string. I still think the name `stamp` isn't super informative. The fact that you have to explain here that stamp represents a "version" shows the name can be improved.

File: service/history/timer_queue_active_task_executor.go (hunk 2 above)
Comment (bergundy) [reply service/history/timer_queue_active_task_executor.go:577]: Yes.

File: service/history/timer_queue_standby_task_executor.go (hunk 3 above)
Comment (bergundy) [reply service/history/timer_queue_standby_task_executor.go:324]: And the transfer task.

File: service/history/workflow/mutable_state_impl.go [hunk 13]
```diff
@@ -4912,6 +4911,35 @@ func (ms *MutableStateImpl) RecordLastActivityStarted(ai *persistencespb.Activit
...
+	// there are two possible cases:
+	//  * this is the first time activity was scheduled
+	//     * in this case we should use current schedule time
+	//	* this is a retry
+	//     * we should recalculate next scheduled time, based on the last time when activity was completed
```
Comment (bergundy) [inline service/history/workflow/mutable_state_impl.go:4919]: What if the activity specified the next retry interval to bypass the retry policy?

File: service/history/workflow/mutable_state_impl.go [hunk 14]
```diff
@@ -4912,6 +4911,35 @@ func (ms *MutableStateImpl) RecordLastActivityStarted(ai *persistencespb.Activit
 	})
 }
 
+func (ms *MutableStateImpl) RegenerateActivityRetryTask(ai *persistencespb.ActivityInfo) error {
+	// there are two possible cases:
+	//  * this is the first time activity was scheduled
+	//     * in this case we should use current schedule time
+	//	* this is a retry
```
Comment (bergundy) [inline service/history/workflow/mutable_state_impl.go:4918]:
```suggestion
	//  * this is a retry
```

File: proto/internal/temporal/server/api/persistence/v1/tasks.proto (hunk 9 above)
Comment (bergundy) [reply proto/internal/temporal/server/api/persistence/v1/tasks.proto:54]: The name stamp is too generic for a task that definition that is fairly broad. I'd call this `conflict_token` (here and everywhere else).

**Review by bergundy — COMMENTED** [review]
Had some comments. I think you should verify that all active task executor and RPC logic is sound in this PR and focus on replication and standby task processing in a follow up PR.

File: service/history/timer_queue_active_task_executor.go (hunk 2 above)
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:577]: func (t *transferQueueTaskExecutorBase) pushActivity(
(and all usagfes of AddActivityTaskRequest)

File: common/serviceerror/activity_stamp_mismatch.go (hunk 6 above)
Comment (ychebotarev) [reply common/serviceerror/activity_stamp_mismatch.go:35]: I do, I want to differentiate between NotFound and ActivityStampMismatch in matching when processing. There is already code path for NotFound, but it is generic.

File: service/history/replication/raw_task_converter.go (hunk 5 above)
Comment (ychebotarev) [reply service/history/replication/raw_task_converter.go:198]: I think I update everything .

File: service/history/timer_queue_active_task_executor.go
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:1]: per offline conversation - check in processSingleActivityTimeoutTask is removed

File: service/history/timer_queue_active_task_executor.go (hunk 2 above)
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:571]: done

File: service/history/workflow/mutable_state_impl.go (hunk 4 above)
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:4918]: The goal is to (re)generate retry task. I don't have a better name for that, but open for proposal

File: common/serviceerror/activity_stamp_mismatch.go (hunk 6 above)
Comment (bergundy) [reply common/serviceerror/activity_stamp_mismatch.go:35]: Why is the error message not good enough?

File: proto/internal/temporal/server/api/historyservice/v1/request_response.proto (hunk 8 above)
Comment (ychebotarev) [reply proto/internal/temporal/server/api/historyservice/v1/request_response.proto:277]: done

File: common/serviceerror/activity_stamp_mismatch.go (hunk 7 above)
Comment (ychebotarev) [reply common/serviceerror/activity_stamp_mismatch.go:58]: ok will change to NotFound

File: service/history/timer_queue_standby_task_executor.go (hunk 3 above)
Comment (ychebotarev) [reply service/history/timer_queue_standby_task_executor.go:324]: done for pushActivity, for standby and transfer.

File: proto/internal/temporal/server/api/persistence/v1/tasks.proto (hunk 9 above)
Comment (ychebotarev) [reply proto/internal/temporal/server/api/persistence/v1/tasks.proto:54]: Will discuss offline, so far I like `stamp` better. I also change the description.

File: proto/internal/temporal/server/api/replication/v1/message.proto (hunk 10 above)
Comment (ychebotarev) [reply proto/internal/temporal/server/api/replication/v1/message.proto:168]: done

File: proto/internal/temporal/server/api/replication/v1/message.proto (hunk 11 above)
Comment (ychebotarev) [reply proto/internal/temporal/server/api/replication/v1/message.proto:170]: done

File: service/history/workflow/mutable_state_impl.go (hunk 13 above)
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:4923]: If you mean "what will happen if retry delay was a part of the failure response" - new scheduled time will be calculated from the new retry policy. If users want to extend retry delay they can do that by specifying new retry policy.

File: service/history/workflow/mutable_state_impl.go (hunk 14 above)
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:4922]: done

File: proto/internal/temporal/server/api/replication/v1/message.proto (hunk 12 above)
Comment (ychebotarev) [reply proto/internal/temporal/server/api/replication/v1/message.proto]: talked offline. will stick to `stamp` for now.

File: service/history/timer_queue_standby_task_executor.go (hunk 3 above)
Comment (ychebotarev) [reply service/history/timer_queue_standby_task_executor.go:324]: keeping this open in case there is something else

File: common/serviceerror/activity_stamp_mismatch.go (hunk 6 above)
Comment (ychebotarev) [reply common/serviceerror/activity_stamp_mismatch.go:35]: actually ok. I will remove it. It will be searchable in the logs either way.

Comment (ychebotarev) [reply common/serviceerror/activity_stamp_mismatch.go]: done

File: proto/internal/temporal/server/api/persistence/v1/tasks.proto (hunk 9 above)
Comment (ychebotarev) [reply proto/internal/temporal/server/api/persistence/v1/tasks.proto:54]: keeping this open in case we come with a better name

Comment (bergundy) [reply proto/internal/temporal/server/api/persistence/v1/tasks.proto:54]: I would at least take the suggested documentation change.

File: proto/internal/temporal/server/api/replication/v1/message.proto [hunk 15]
```diff
@@ -165,6 +165,12 @@ message SyncActivityTaskAttributes {
     string last_started_build_id = 16;
     // workflows redirect_counter value when this activity started last time
     int64 last_started_redirect_counter = 17;
+    // The first time the activity was scheduled.
+    google.protobuf.Timestamp first_scheduled_time = 18;
+    // The last time an activity attempt completion was recorded by the server.
+    google.protobuf.Timestamp last_attempt_complete_time = 19;
+    // Stamp represents the internal “version” of the activity options and can/will be changed with Activity API.
```
Comment (bergundy) [inline proto/internal/temporal/server/api/replication/v1/message.proto:172]:
```suggestion
    // Stamp represents the internal "version" of the the activity options. It is monotonically increments when the activity options are changed via an API request.
```

File: proto/internal/temporal/server/api/errordetails/v1/message.proto [hunk 16]
```diff
@@ -67,3 +67,7 @@ message StickyWorkerUnavailableFailure {
 
 message ObsoleteDispatchBuildIdFailure {
 }
+
+message ActivityStampMismatchFailure {
+
+}
```
Comment (bergundy) [inline proto/internal/temporal/server/api/errordetails/v1/message.proto:73]: This can be deleted now.

File: service/history/timer_queue_active_task_executor.go [hunk 17]
```diff
@@ -230,19 +230,13 @@ func (t *timerQueueActiveTaskExecutor) executeActivityTimeoutTask(
...
 	if !ok {
 		// if activity is not found, the timer is invalid.
 		return nil
 	}
 
-	if task.Stamp != ai.Stamp {
-		// if this task is invalid - we want to generate new activity timeout task
-		// it will be done in closeTransactionPrepareTasks call
-		return weContext.UpdateWorkflowExecutionAsActive(ctx, t.shardContext)
-	}
```
Comment (bergundy) [inline service/history/timer_queue_active_task_executor.go:244]: Ah this makes sense. So basically when you `LoadAndSortActivityTimers()` you'll generate only relevant timer tasks.

Maybe put a comment in this function explaining that there's no need to check the stamp here since we generate the timer sequence on the fly and it is always valid given the current activity state.

File: service/history/workflow/mutable_state_impl.go (hunk 13 above)
Comment (bergundy) [reply service/history/workflow/mutable_state_impl.go]: I see... so maybe this should be clarified in a comment. Ideally you'll move this whole comment (including my suggestion) into the function's docstring where it'll be more discoverable.

File: service/history/workflow/mutable_state_impl.go [hunk 18]
```diff
@@ -6479,6 +6511,8 @@ func (ms *MutableStateImpl) ShouldResetActivityTimerTaskMask(current, incoming *
 		return true
 	} else if current.Attempt != incoming.Attempt {
 		return true
+	} else if current.Stamp != incoming.Stamp {
```
Comment (bergundy) [inline service/history/workflow/mutable_state_impl.go:6514]: Update the comment above please.

**Review by bergundy — APPROVED** [review]
Approved with some comments. Mostly around documentation but also wondering if we have enough test coverage for this feature.

File: proto/internal/temporal/server/api/errordetails/v1/message.proto (hunk 16 above)
Comment (ychebotarev) [reply proto/internal/temporal/server/api/errordetails/v1/message.proto:73]: done

File: proto/internal/temporal/server/api/replication/v1/message.proto (hunk 15 above)
Comment (ychebotarev) [reply proto/internal/temporal/server/api/replication/v1/message.proto:172]: update comments.

File: service/history/workflow/mutable_state_impl.go (hunk 18 above)
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:6514]: done

File: service/history/timer_queue_active_task_executor.go (hunk 17 above)
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:244]: add a comment

File: proto/internal/temporal/server/api/persistence/v1/tasks.proto (hunk 9 above)
Comment (ychebotarev) [reply proto/internal/temporal/server/api/persistence/v1/tasks.proto:54]: I added a more detailed comments .

File: service/history/workflow/mutable_state_impl.go (hunk 13 above)
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go]: the comment clearly state how schedule time will be recalculated. But I also add a note.

File: service/history/timer_queue_active_task_executor.go [hunk 19]
```diff
@@ -230,19 +230,13 @@ func (t *timerQueueActiveTaskExecutor) executeActivityTimeoutTask(
 	if mutableState == nil || !mutableState.IsWorkflowExecutionRunning() {
 		return nil
 	}
-	ai, ok := mutableState.GetActivityInfo(task.EventID)
+	_, ok := mutableState.GetActivityInfo(task.EventID)
 
 	if !ok {
 		// if activity is not found, the timer is invalid.
 		return nil
 	}
```
Comment (yycptt) [inline service/history/timer_queue_active_task_executor.go:238]: I think we agreed to remove this check as well per our offline discussion since the task is not for a specific activity. On L269 we will check if activity exists based on the actual activity timers.

Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:238]: we agree to remove stamp check. But you right - this one is not needed for the same reasons. I will remove it.

**Review by yycptt — APPROVED** [review]
